### ISSUE TITLE
Opaque type completely erased

### ISSUE THREAD

**Issue by jsgf** [servo/rust-bindgen#386]
If I have the input:
```
#include <string>

extern std::string *into_cpp(const char *cstr, size_t len);
extern size_t str_len(const std::string *str);
```

And generate a binding with:
```
bindgen mystring.h --enable-cxx-namespaces --generate types,methods,functions --whitelist-function=into_cpp --whitelist-function=str_len --whitelist-type=str::string '--opaque-type=.*' -- -x c++ -std=gnu++14
```

I get the output:

```
/* automatically generated by rust-bindgen */

pub mod root {
    #[allow(unused_imports)]
    use self::super::root;
    pub mod std {
        #[allow(unused_imports)]
        use self::super::super::root;
        pub mod namespace {
            #[allow(unused_imports)]
            use self::super::super::super::root;
        }
    }
    pub mod __gnu_cxx {
        #[allow(unused_imports)]
        use self::super::super::root;
    }
    extern "C" {
        #[link_name = "_Z8into_cppPKcm"]
        pub fn into_cpp(cstr: *const ::std::os::raw::c_char, len: u64)
         -> *mut [u64; 3usize];
    }
    extern "C" {
        #[link_name =
              "_Z7str_lenPKSt14basic_fbstringIcSt11char_traitsIcESaIcESt13fbstring_coreIcEE"]
        pub fn str_len(str: *const [u64; 3usize]) -> u64;
    }
}
```

`std::string` is completely erased, replaced with `[u64; 3usize]`, and not even a `pub type ...` left.

(Hm, on Macos using the Brew version of clang, it segfaults with what appears to be infinite recursion.)

**Comment by emilio** [comment]
This is sort of expected, we use `opaque` to replace layouts we can't handle.

For example, if we have a type like:

```c++
template<typename ...Args>
struct Baz {
    // Something we can't handle with Args...
};

struct Foo {
    Baz<i32, Foo> whatever;
};
```

The only way to get a correct layout is making `Baz` opaque, in which case we'll detect it's a specialization of an opaque type we know the layout and alignment of, and generate a bunch of bytes to ensure the correct layout for `Foo`.

This should be better documented though.

The solution to hide types is general to use `--blacklist-type`.

**Comment by jsgf** [comment]
Isn't this backwards? `--blacklist-type` adds the pattern to `options.hide_type`, but `--opaque-type` goes into `options.opaque_type`.

**Comment by emilio** [comment]
Yes, blacklist does that, but opaque doesn't do that when there's template parameters for said reason... (note that `std::string` is really a subclass with template params).

Though you're right we should improve that actually.

We should only fall back to that behavior if the type is opaque _and_ it's layout is unknown (and potentially _and_ if referencing it would incur in no applicable template arguments).

Will see if I can come up with something.

**Comment by emilio** [comment]
Actually my comment may be misleading, we may not have specific code for template parameters.

**Comment by jsgf** [comment]
Sometimes I can get it to produce `pub type string = [u64; 3usize]`, but I forgot to note the options and I haven't managed to get it back.

**Comment by emilio** [comment]
Hmm... Yeah, that'd be possible. I think right now we stop tracing opaque types, which would make string not appear, but I guess what we want is not tracing the children.

That'd explain that.

**Comment by emilio** [comment]
sorry, got sidetracked, I think I know what this is.

**Comment by emilio** [comment]
#387 should do it :)

**Comment by emilio** [comment]
Then there's the code generated for strings, which needs to be tweaked too. This may be a bit more complex because there are chances to break stuff, so it'll take a bit more time to test it properly.

**Comment by emilio** [comment]
Mainly, I don't think we should go ahead and generate a placeholder when we know the layout of that opaque type, which is the problem the placeholder is trying to solve.

In order to do that, we can as an initial step just chec for template parameters in the type, which would be ok for this case because it's an alias, and that could not break stuff I think (because we should know the layout of everything non-dependent on template params).

**Comment by emilio** [comment]
Added a few more fixes to #387 that should also fix the `pub mod namespace` thing (modern stdlibs are using `inline namespace`s, that we didn't handle, and makes your example work as you expect for me.

**Comment by emilio** [comment]
This should be fixed now, thanks for reporting it!
